// ngx-lazy-load-images/esm2015/lib/ngx-lazy-load-images.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Renderer2, Input, NgZone, Inject, PLATFORM_ID, OnInit, OnDestroy } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
/**
 * Angular Lazy Loading Images Directive
 *
 * The library allows to lazy load images from your web application
 * using the MutationObserver and the IntersectionObserver. Images will be loaded as
 * soon as they enter the viewport in a non-blocking way.
 */
let LazyLoadImagesDirective = class LazyLoadImagesDirective {
    constructor(element, renderer, ngZone, platformId) {
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.platformId = platformId;
        this.rootElement = element.nativeElement;
    }
    init() {
        this.registerIntersectionObserver();
        this.observeDOMChanges(this.rootElement, () => {
            const imagesFoundInDOM = this.getAllImagesToLazyLoad(this.rootElement);
            imagesFoundInDOM.forEach((image) => this.intersectionObserver.observe(image));
        });
    }
    ngOnInit() {
        if (!this.isBrowser()) {
            return;
        }
        require('intersection-observer');
        this.ngZone.runOutsideAngular(() => this.init());
    }
    ngOnDestroy() {
        if (this.intersectionObserver) {
            this.intersectionObserver.disconnect();
        }
    }
    isBrowser() {
        return isPlatformBrowser(this.platformId);
    }
    registerIntersectionObserver() {
        this.intersectionObserver = new IntersectionObserver(images => images.forEach(image => this.onIntersectionChange(image)), this.intersectionObserverConfig instanceof Object ? this.intersectionObserverConfig : undefined);
        return this.intersectionObserver;
    }
    observeDOMChanges(rootElement, onChange) {
        // Create a Mutation Observer instance
        const observer = new MutationObserver(() => onChange());
        // Observer Configuration
        const observerConfig = {
            attributes: true,
            characterData: true,
            childList: true,
            subtree: true
        };
        // Observe Directive DOM Node
        observer.observe(rootElement, observerConfig);
        // Fire onChange callback to check current DOM nodes
        onChange();
        return observer;
    }
    getAllImagesToLazyLoad(pageNode) {
        return Array.from(pageNode.querySelectorAll('img[data-src], [data-srcset], [data-background-src]'));
    }
    onIntersectionChange(image) {
        if (!image.isIntersecting) {
            return;
        }
        this.onImageAppearsInViewport(image.target);
    }
    onImageAppearsInViewport(image) {
        if (image.dataset.src) {
            this.renderer.setAttribute(image, 'src', image.dataset.src);
            this.renderer.removeAttribute(image, 'data-src');
        }
        if (image.dataset.srcset) {
            this.renderer.setAttribute(image, 'srcset', image.dataset.srcset);
            this.renderer.removeAttribute(image, 'data-srcset');
        }
        if (image.dataset.backgroundSrc) {
            this.renderer.setStyle(image, 'background-image', `url(${image.dataset.backgroundSrc})`);
            this.renderer.removeAttribute(image, 'data-background-src');
        }
        // Stop observing the current target
        if (this.intersectionObserver) {
            this.intersectionObserver.unobserve(image);
        }
    }
};
LazyLoadImagesDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: NgZone },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
__decorate([
    Input('lazy-load-images')
], LazyLoadImagesDirective.prototype, "intersectionObserverConfig", void 0);
LazyLoadImagesDirective = __decorate([
    Directive({
        selector: '[lazy-load-images]'
    }),
    __param(3, Inject(PLATFORM_ID))
], LazyLoadImagesDirective);
export { LazyLoadImagesDirective };
